src/aosm/azext_aosm/common/registry.py

```diff
@@ -0,0 +1,469 @@
+                    "credentials scoped only to the store. This requires Docker to be "
+                    "installed locally."
+                ) from error
+
+            # Otherwise, the most likely failure is that the image already exists in the artifact
+            # store, so don't fail at this stage, log the error.
+            logger.warning(
+                (
+                    "Failed to copy %s from %s to %s. If this failure is because it already "
+                    "exists in the target registry we can continue. If the failure was for "
+                    "another reason, for example it does not exist in the source registry, "
+                    "this is likely fatal. Attempting to continue.\n"
+                    "%s"
+                ),
+                source_image,
+                self.registry_name,
+                target_acr,
+                error,
+            )
+
+    def find_image(
+        self, image: str, version: str
+    ) -> Union[Tuple["AzureContainerRegistry", str], Tuple[None, None]]:
+        """
+        Find whether the given image exists in this registry.
+
+        :param image: The image to find in the registry
+            (the image should be stripped of all namespace)
+        :param version: The version of the image (also known as image tag)
+        :return: The registry and namespace for the image
+        """
+        for namespace in self.registry_namespaces:
+            image_with_namespace = f"{namespace}{image}:{version}"
+            try:
+                acr_source_get_images_cmd = [
+                    str(shutil.which("az")),
+                    "acr",
+                    "repository",
+                    "show",
+                    "--name",
+                    self.registry_name,
+                    "--image",
+                    image_with_namespace,
+                ]
+
+                call_subprocess_raise_output(acr_source_get_images_cmd)
+
+                return (self, namespace)
+
+            except CLIError as error:
+                logger.debug(
+                    ("Image %s, version %s not found in %s registry."),
+                    image_with_namespace,
+                    version,
+                    self.registry_name,
+                )
+                logger.debug(error, exc_info=True)
+
+        return None, None
+
+
+class ContainerRegistryHandler:
+    """
+    A class to handle all the registries and their images.
+    """
+
+    def __init__(self, image_sources):
+        self.image_sources = image_sources
+        self.registry_list = self._create_registry_list()
+
+    def _create_registry_list(self) -> List[ContainerRegistry]:
+        """
+        Create a list of registry objects from the registries provided by the user.
+
+        :return: A list of registry objects
+        """
+
+        registry_name_to_class_dict: Dict[str, ContainerRegistry] = {}
```

Comment:
This renaming has been really helpful for my understanding of what this method is doing :-)

It's now made me wonder if a better way of doing this would be to loop over the objects in `registry_object_list` and check if an `object.registry_name` is already there. I'd argue this is clearer, and does away with the `registry_name_to_object_dict` variable altogether.

Just a suggestion, but here's how the code could look (obvs not tested!):

```
        registry_object_list: List[ContainerRegistry] = []
        logger.debug("Creating registry list from the provided registries")

        for registry in self.image_sources:
            registry = clean_registry_name(registry)

            parts = registry.split("/", 1)
            registry_name = parts[0]
            registry_namespace = parts[1] if len(parts) > 1 else None

            if registry_namespace:
                if not registry_namespace.endswith("/"):
                    registry_namespace += "/"
            else:
                registry_namespace = ""

            # Get the registry_object with matching registry_name from the list, or None if not found
            reg_object = next((reg_object for reg_object in registry_object_list if reg_object.registry_name == registry_name), None)
            if reg_object:
                # Object already exists, add namespace to it
                reg_object.add_namespace(registry_namespace)
            else:
                # Create a new registry object
                acr_match = re.match(ACR_REGISTRY_NAME_PATTERN, registry_name)
                if acr_match:
                    registry_object_list.append(AzureContainerRegistry(registry_name))
                else:
                    registry_object_list.append(UniversalRegistry(registry_name))
                registry_object_list[-1].add_namespace(registry_namespace)

        return registry_object_list
```